client.ts

```diff
@@ -0,0 +1,231 @@
+      port: options.port || "3000",
+      reconnect: options.reconnect || true,
+      protocol: options.protocol || "ws",
+    };
+    this.decoder = new TextDecoder();
+    this.reconnectCount = 0;
+    this.listening_to = {};
+    this.message_queue = [];
+    this.ready = true;
+
+    this._connectToSocketServer();
+  }
+
+  // FILE MARKER - METHODS FOR PUBLIC USE //////////////////////////////////////////////////////////
+
+  /**
+   * @description
+   *     On receipt of a message in the following channel or on the following event, perform the
+   *     specified callback.
+   *
+   * @param {String} channelOrEvent
+   *     The name of the channel or event.
+   * @param {Function} callback
+   *     The callback to execute on receipt of a message from the channel or event.
+   */
+  public on(channelOrEvent: string, callback) {
+    if (this._isClientReady()) {
+      if (!this.listening_to[channelOrEvent]) {
+        this.listening_to[channelOrEvent] = null;
+      }
+      this.listening_to[channelOrEvent] = callback;
+      const message = JSON.stringify({ listening_to: channelOrEvent });
+      const encoded = new TextEncoder().encode(message);
+      this.message_queue.push(encoded);
+      this._sendMessagesToSocketServer();
+    } else {
+      setTimeout(() => this.on(channelOrEvent, callback), 5);
+    }
+  }
+
+  /**
+   * @description
+   *     Send a message to a channel or an event.
+   *
+   * @param {String} channelOrEvent
+   *     The name of the channel or event.
+   * @param {String} message
+   *     The message to send to the channel or event.
+   */
+  public to(channelOrEvent: string, message: {[key: string]: unknown} | Event): void {
+    if (channelOrEvent) {
+      const messageString = JSON.stringify({ [channelOrEvent]: message });
+      const encodedMessage = new TextEncoder().encode(messageString);
+      this.message_queue.push(encodedMessage);
+      this._sendMessagesToSocketServer();
+    }
+  }
+
+  // FILE MARKER - METHODS FOR INTERNAL USE ////////////////////////////////////////////////////////
+
+  /**
+   * Check if connection is ready for events.
+   */
+  private _isClientReady(): boolean {
+    return this.connection.readyState === 1;
+  }
+
+  private _reconnectSuccessful(previousId: number): void {
+    this.reconnectCount += 1;
+    // server can react if needed to this connection id
+    this.to("reconnect", {
+      previousId,
+      //id: this.connection.id,
+      reconnectCount: this.reconnectCount,
+    });
+  }
+
+  /**
+   * Connect to the socket server at the hostname and port specified in the configs.
+   */
+  private _connectToSocketServer(reconnect?: boolean) {
+    //const previousId = reconnect ? this.connection.id : null;
+
+    this.connection = new WebSocket(
+      `${this.configs.protocol}://${this.configs.hostname}:${this.configs.port}`,
+    );
+
+    //if (previousId) {
+    //  this._reconnectSuccessful(previousId);
+    //}
```

Comment:
Only true changes i made are commenting these lines out - `id` doesn't exist on websocket so i assume this didn't work before?